e2e/nomostest/clean.go

```diff
@@ -126,6 +133,95 @@ func Clean(nt *NT) error {
+	}
+	if err := testpredicates.StatusEquals(nt.Scheme, kstatus.CurrentStatus)(rmObj); err != nil {
+		return fmt.Errorf("reconciler-manager not healthy: %v", err)
+	}
+	// If reconciler-manager is healthy, uninstall all the unmanaged RootSyncs & RepoSyncs.
+	// If deletion-propagation is enabled, this will also recursively delete all
+	// managed objects, including other managed packages.
+	if err := uninstallUnmanagedPackagesOfType(nt, kinds.RepoSyncV1Beta1()); err != nil {
+		return fmt.Errorf("uninstalling unmanged RepoSyncs: %v", err)
+	}
+	if err := uninstallUnmanagedPackagesOfType(nt, kinds.RootSyncV1Beta1()); err != nil {
+		return fmt.Errorf("uninstalling unmanged RootSyncs: %v", err)
+	}
+	return nil
+}
+
+func uninstallUnmanagedPackagesOfType(nt *NT, gvk schema.GroupVersionKind) error {
+	// Find all packages
+	nt.T.Logf("[CLEANUP] Listing %s objects...", gvk.Kind)
+	rsObjList := &unstructured.UnstructuredList{}
+	rsObjList.SetGroupVersionKind(kinds.ListGVKForItemGVK(gvk))
+	if err := nt.KubeClient.List(rsObjList, client.InNamespace(configmanagement.ControllerNamespace)); err != nil {
+		return fmt.Errorf("listing %s: %v", gvk.Kind, err)
+	}
+	// Find the unmanaged packages
+	var rsObjs []*unstructured.Unstructured
+	for _, item := range rsObjList.Items {
+		if core.GetAnnotation(&item, metadata.ResourceManagementKey) != metadata.ResourceManagementEnabled {
+			rsObjs = append(rsObjs, &item)
+		} else {
+			nt.T.Logf("[CLEANUP] Skipping deletion of managed %s object %s",
+				gvk.Kind, client.ObjectKeyFromObject(&item))
+		}
+	}
+	// Delete the unmanaged packages in parallel
+	nt.T.Log("[CLEANUP] Deleting unmanaged %s objects...", gvk.Kind)
+	tg := taskgroup.New()
+	for _, obj := range rsObjs {
+		nn := client.ObjectKeyFromObject(obj)
+		if !obj.GetDeletionTimestamp().IsZero() {
+			nt.T.Logf("[CLEANUP] Already terminating %s object %s ...", gvk.Kind, nn)
+		} else {
+			nt.T.Logf("[CLEANUP] Deleting %s object %s ...", gvk.Kind, nn)
+			if err := nt.KubeClient.Delete(obj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+				if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+					// skip waiting
+					nt.T.Logf("[CLEANUP] Confirmed removal of %s object %s", gvk.Kind, nn)
+					continue
+				}
+				// return an error, but don't prevent other deletes or waits
+				tg.Go(func() error {
+					return fmt.Errorf("unable to delete %s object %s: %w",
+						gvk.Kind, nn, err)
+				})
+				continue
+			}
+		}
+		tg.Go(func() error {
+			nt.T.Logf("[CLEANUP] Waiting for removal or failure of %s object %s ...", gvk.Kind, nn)
+			err := nt.Watcher.WatchObject(gvk, nn.Name, nn.Namespace,
+				[]testpredicates.Predicate{
+					testpredicates.Or(nt.Logger,
+						testpredicates.ObjectNotFoundPredicate(nt.Scheme),
+						testpredicates.StatusEquals(nt.Scheme, kstatus.FailedStatus),
+						testpredicates.HasConditionStatus(nt.Scheme,
+							string(v1beta1.RootSyncReconcilerFinalizerFailure), corev1.ConditionTrue),
```

Comment:
This string should probably change based on whether the GVK is RootSync or RepoSync. It happens that the `RootSyncReconcilerFinalizerFailure`/`RepoSyncReconcilerFinalizerFailure` constants currently have the same value, but I don't think this should be assumed given they are separate constants. Alternatively, we could consolidate the constants to something like `ReconcilerFinalizerFailure` if the intent is for them to have the same value.

Reply:
Pulled out the conditions above the loop and added a switch for this.

Reply:
Looks better, thanks